File: src/filter.ts
import {TLog} from './log';

export type Filter = {
	applied: boolean;
	level: {
		debug: boolean;
		warning: boolean;
		info: boolean;
	};
	type: {
		enter: boolean;
		exit: boolean;
	};
	prefix: string;
	taskname: string;
	syscallname: string;
};

export const defaultFilter: Filter = {
	applied: false,
	level: {debug: true, warning: true, info: true},
	type: {enter: true, exit: true},
	prefix: '',
	taskname: '',
	syscallname: '',
};

const regexIsValid = (regex: string) => {
	try {
		new RegExp(regex);
		return true;
	} catch (error) {
		console.log('cant parse regex, ignore');
		return false;
	}
};

export function checkSuggest(log: TLog, filter: Filter) {
	// TODO not parse regex every check
	if (filter.prefix && regexIsValid(filter.prefix)) {
		const prefixRegex = new RegExp(filter.prefix);
		if (!prefixRegex.test(log.LogPrefix)) return false;
	}
	if (filter.taskname && regexIsValid(filter.taskname)) {
		const tasknameRegex = new RegExp(filter.taskname);
		if (!tasknameRegex.test(log.Taskname)) return false;
	}
	if (filter.syscallname && regexIsValid(filter.syscallname)) {
		const syscallnameRegex = new RegExp(filter.syscallname);
		if (!syscallnameRegex.test(log.Syscallname)) return false;
	}

	if (log.LogType == 'E') {
		if (!filter.type.enter) return false;
	} else {
		if (!filter.type.exit) return false;
	}

	if (log.level == 'info') {
		if (!filter.level.info) return false;
	} else if (log.level == 'debug') {
		if (!filter.level.debug) return false;
	} else if (log.level == 'warning') {
		if (!filter.level.warning) return false;
	}

	return true;
}

File: filter.go
package main

import (
	"fmt"
	"regexp"
)

type FilterDto struct {
	Applied bool
	Level   struct {
		Debug   bool
		Warning bool
		Info    bool
	}
	Type struct {
		Enter bool
		Exit  bool
	}
	Prefix      string
	Taskname    string
	Syscallname string
}

type TFilter struct {
	Applied     bool
	Levels      []string
	Types       []string
	Prefix      regexp.Regexp
	Taskname    regexp.Regexp
	Syscallname regexp.Regexp
}

var defaultFilter = TFilter{
	Applied:     true,
	Levels:      []string{"debug", "info", "warning"},
	Types:       []string{"E", "X"},
	Prefix:      *regexp.MustCompile(""),
	Taskname:    *regexp.MustCompile(""),
	Syscallname: *regexp.MustCompile(""),
}

func regexIsValid(regex string) bool {
	_, err := regexp.Compile(regex)
	if err != nil {
		fmt.Println("can't parse regex, ignore")
		return false
	}

	return true
}

func toTFilter(dto FilterDto) TFilter {
	levels := []string{}
	types := []string{}

	if dto.Level.Info {
		levels = append(levels, "info")
	}
	if dto.Level.Debug {
		levels = append(levels, "debug")
	}
	if dto.Level.Warning {
		levels = append(levels, "warning")
	}
	if dto.Type.Exit {
		types = append(types, "X")
	}
	if dto.Type.Enter {
		types = append(types, "E")
	}
	if len(levels) == 0 {
		levels = []string{"info", "debug", "warning"}
	}
	if len(types) == 0 {
		types = []string{"E", "X"}
	}

	return TFilter{
		Applied:     dto.Applied,
		Levels:      levels,
		Types:       types,
		Prefix:      *regexp.MustCompile(dto.Prefix),
		Taskname:    *regexp.MustCompile(dto.Taskname),
		Syscallname: *regexp.MustCompile(dto.Syscallname),
	}
}

func checkSuggest(log TLog, filter TFilter) bool {
	return contains(filter.Levels, log.Level) &&
		contains(filter.Types, log.LogType) &&
		filter.Prefix.MatchString(log.LogPrefix) &&
		filter.Taskname.MatchString(log.Taskname) &&
		filter.Syscallname.MatchString(log.Syscallname)
}

func contains(arr []string, val string) bool {
	for _, item := range arr {
		if item == val {
			return true
		}
	}

	return false
}
